src/main/scala/diplomacy/LazyModule.scala

```diff
@@ -71,217 +121,386 @@ abstract class LazyModule()(implicit val p: Parameters)
-  // Otherwise, anonymous children (Monitors for example) will not be clocked
-  val childClock = Wire(Clock())
-  val childReset = Wire(Reset())
+  // These wires are the default clock+reset for all LazyModule children.
+  // It is recommended to drive these even if you manually drive the [[clock]] and [[reset]] of all of the
+  // [[LazyRawModuleImp]] children.
+  // Otherwise, anonymous children ([[Monitor]]s for example) will not have their [[clock]] and/or [[reset]] driven properly.
+  /** drive clock explicitly. */
+  val childClock: Clock = Wire(Clock())
+  /** drive reset explicitly. */
+  val childReset: Reset = Wire(Reset())
+  // the default is that these are disabled
   childClock := Bool(false).asClock
   childReset := chisel3.DontCare
   val (auto, dangles) = withClockAndReset(childClock, childReset) {
     instantiate()
   }
 }
 
-class SimpleLazyModule(implicit p: Parameters) extends LazyModule
-{
+/** Used for a [[LazyModule]] which does not need to define any [[LazyModuleImp]] implementation.
+  *
+  * It can be used as wrapper that only instantiates and connects [[LazyModule]]s.
+  */
+class SimpleLazyModule(implicit p: Parameters) extends LazyModule {
   lazy val module = new LazyModuleImp(this)
 }
 
-trait LazyScope
-{
+/** Allows dynamic creation of [[Module]] hierarchy and "shoving" logic into a [[LazyModule]]. */
+trait LazyScope {
   this: LazyModule =>
   override def toString: String = s"LazyScope named $name"
-  def apply[T](body: => T) = {
+
+  /** Evaluate `body` in the current [[LazyModule.scope]] */
+  def apply[T](body: => T): T = {
+    // Preserve the previous value of the [[LazyModule.scope]], because when calling [[apply]] function,
+    // [[LazyModule.scope]] will be altered.
     val saved = LazyModule.scope
+    // [[LazyModule.scope]] stack push.
     LazyModule.scope = Some(this)
+    // Evaluate [[body]] in the current `scope`, saving the result to [[out]].
     val out = body
-    require (LazyModule.scope.isDefined, s"LazyScope ${name} tried to exit, but scope was empty!")
-    require (LazyModule.scope.get eq this, s"LazyScope ${name} exited before LazyModule ${LazyModule.scope.get.name} was closed")
+    // Check that the `scope` after evaluating `body` is the same as when we started.
+    require(LazyModule.scope.isDefined, s"LazyScope $name tried to exit, but scope was empty!")
+    require(LazyModule.scope.get eq this, s"LazyScope $name exited before LazyModule ${LazyModule.scope.get.name} was closed")
+    // [[LazyModule.scope]] stack pop.
     LazyModule.scope = saved
     out
   }
 }
 
-object LazyScope
-{
+/** Used to automatically create a level of module hierarchy (a [[SimpleLazyModule]]) within which [[LazyModule]]s can be instantiated and connected.
+  *
+  * It will instantiate a [[SimpleLazyModule]] to manage evaluation of `body`.
+  * And evaluate `body` codes in this scope.
+  */
+object LazyScope {
+  /** Create a [[LazyScope]] with name in its val name.
```

Comment:
```suggestion
  /** Create a [[LazyScope]].
```

I don't really know what "with name in its val name" means and it isn't very helpful. I'd suggest deleting this or rewriting it to be more clear.